src/components/periodic-report/periodic-report.service.ts

```diff
@@ -0,0 +1,423 @@
+import {
+  CreatePeriodicReport,
+  IPeriodicReport,
+  PeriodicReport,
+  PeriodicReportListInput,
+  ReportType,
+  SecuredPeriodicReportList,
+} from './dto';
+
+@Injectable()
+export class PeriodicReportService {
+  private readonly securedProperties = {
+    type: true,
+    start: true,
+    end: true,
+    reportFile: true,
+  };
+
+  constructor(
+    private readonly db: DatabaseService,
+    private readonly files: FileService,
+    @Logger('periodic:report:service') private readonly logger: ILogger,
+    @Inject(forwardRef(() => AuthorizationService))
+    private readonly authorizationService: AuthorizationService
+  ) {}
+
+  @OnIndex()
+  async createIndexes() {
+    return [
+      'CREATE CONSTRAINT ON (n:PeriodicReport) ASSERT EXISTS(n.id)',
+      'CREATE CONSTRAINT ON (n:PeriodicReport) ASSERT n.id IS UNIQUE',
+    ];
+  }
+
+  async create(
+    input: CreatePeriodicReport,
+    session: Session
+  ): Promise<PeriodicReport> {
+    const id = await generateId();
+    const createdAt = DateTime.local();
+
+    try {
+      const createPeriodicReport = this.db
+        .query()
+        .create([
+          [
+            node(
+              'newPeriodicReport',
+              ['PeriodicReport', 'BaseNode', `${input.type}Report`],
+              {
+                createdAt,
+                id,
+              }
+            ),
+          ],
+          ...property('type', input.type, 'newPeriodicReport'),
+          ...property('start', input.start, 'newPeriodicReport'),
+          ...property('end', input.end, 'newPeriodicReport'),
+        ])
+        .return('newPeriodicReport.id as id');
+      const result = await createPeriodicReport.first();
+
+      if (!result) {
+        throw new ServerException('Failed to create a periodic report');
+      }
+
+      await this.db
+        .query()
+        .match(
+          node(
+            'node',
+            input.type === ReportType.Progress ? 'Engagement' : 'Project',
+            { id: input.projectOrEngagementId }
+          )
+        )
+        .match(node('periodicReport', 'PeriodicReport', { id: result.id }))
+        .create([
+          node('node'),
+          relation('out', '', 'report', {
+            active: true,
+            createdAt: DateTime.local(),
+          }),
+          node('periodicReport'),
+        ])
+        .run();
+
+      return await this.readOne(id, session);
+    } catch (exception) {
+      throw new ServerException('Could not create periodic report', exception);
+    }
+  }
+
+  async uploadFile(
+    reportId: ID,
+    file: CreateDefinedFileVersionInput,
+    session: Session
+  ) {
+    const reportFileId = (await generateId()) as FileId;
+
+    const report = await this.readOne(reportId, session);
+
+    await this.files.createDefinedFile(
+      reportFileId,
+      report.end.toISODate(),
+      session,
+      reportId,
+      'reportFile',
+      file,
+      'periodicReport.reportFile'
+    );
```

Comment:
I know this is a little confusing, but we actually want to do this on report create. All of our _files_ are actually just holders for _file versions_. By creating this "holder" when the object is created we can always reference the file/holder ID to connect the current upload/file version to.

Does that make sense?
See `growthPlan` in EngagementService for a concrete example you could follow.

Reply:
Nevermind did it myself 37529aaf319c0aad67293f695f1619b2907783ca